server/src/main/java/org/elasticsearch/cluster/node/DiscoveryNodes.java

```diff
@@ -71,7 +77,9 @@ private DiscoveryNodes(
         this.masterNodes = masterNodes;
         this.ingestNodes = ingestNodes;
         this.masterNodeId = masterNodeId;
+        this.masterNode = masterNodeId == null ? null : nodes.get(masterNodeId);
```

Comment:
Could we assert that `(masterNodeId == null) == (masterNode == null)` just to clarify that these things are always consistent?